gha: set default permissions, and add golangci-lint

thaJeztah opened the pull request:
- relates to https://github.com/opencontainers/go-digest/pull/105

thaJeztah committed:
gha: set default permission to read

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
gha: add golangci-lint

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Good, it works! 😄

thaJeztah committed:
gha: add go1.23 - go1.26

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
gha: set GOTOOLCHAIN=local

Prevent implicit upgrades of Go

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

AkihiroSuda commented:
Is this still a draft?

sudo-bmitch commented on .github/workflows/test.yml:
```diff
@@ -1,12 +1,18 @@
     strategy:
       matrix:
-        go-version: [1.18.x, 1.19.x, 1.20.x, 1.21.x, 1.22.x]
+        go-version: [1.18.x, 1.19.x, 1.20.x, 1.21.x, 1.22.x, 1.23.x, 1.24.x, 1.25.x, 1.26.x]
```
Rather than trying to maintain an ever growing list of versions, in image-spec we test stable, oldstable, and whatever is in the go.mod file: https://github.com/opencontainers/image-spec/blob/main/.github/workflows/docs-and-linting.yml

thaJeztah replied:
Yup, I usually pick `oldest` (go.mod), `oldstable`, `stable`, which normally book-ends the versions. This one was already using a list, so kept it, but I'd personally be fine with skipping the intermediates.

That said; it runs in under a minute, so still .. reasonable.

thaJeztah committed:
gha: use go.mod for "minimum" Go version, and use custom name

actions/setup-go falls back to the configured "go-version-file" if
the "go-version" is empty; this allows using the version defined in
go.mod to be used for the minimum supported version.

also specify a custom name instead of the actual go version, so that
branch protection rules don't have to be updated if the go version
changes.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

sudo-bmitch commented on .github/workflows/test.yml:
```diff
@@ -1,21 +1,40 @@
           go-version: ${{ matrix.go-version }}
+          go-version-file: go.mod # used when go-version is empty.
       - name: Checkout code
         uses: actions/checkout@v7
```
Move the checkout step above the go install to fix the GHA error.

thaJeztah replied:
Ha! Yes, I was just updating; was scratching my head and then spotted it :joy:

sudo-bmitch commented on .github/workflows/test.yml:
```diff
@@ -1,21 +1,40 @@
+      - name: Lint Blake3
+        uses: golangci/golangci-lint-action@v9
+        with:
+          version: v2.11
```
I believe this is a duplicate lint as above. According to https://github.com/golangci/golangci-lint-action#working-directory they support `working-directory`:

```suggestion
          version: v2.11
          working-directory: blake3
```

thaJeztah replied:
You're right! Thought I added a path, but ... looks like I forgot; let me fix

thaJeztah commented:
@AkihiroSuda ptal 🤗
